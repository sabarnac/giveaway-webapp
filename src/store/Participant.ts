import { observable, computed } from "mobx";

/**
 * Class containing the details of a participants' avatar.
 */
export class Avatar {
  @observable private _url: string;
  @observable private _altText: string;

  public constructor(url: string, altText: string) {
    this._url = url;
    this._altText = altText;
  }

  /**
   * Gets the avatar URL.
   * @return {string} The avatar URL.
   */
  public get url(): string {
    return this._url;
  }

  /**
   * Gets the avatar image alternate text.
   * @return {string} The avatar image alternate text.
   */
  public get altText(): string {
    return this._altText;
  }

  private isSameUrl = (url1: string, url2: string): boolean => url1 === url2;

  private isSameAltText = (altText1: string, altText2: string): boolean =>
    altText1 === altText2;

  /**
   * Determines whether another avatar is identical to the current one.
   * @param  {Avatar} otherAvatar The other avatar to compare against.
   * @returns {boolean} Whether the other avatar is equal to the current one.
   */
  public equals = (otherAvatar: Avatar): boolean =>
    this.isSameUrl(this._url, otherAvatar._url) &&
    this.isSameAltText(this._altText, otherAvatar._altText);
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  @observable private _name: string;
  @observable private _avatar?: Avatar;

  public constructor(name: string, avatar?: Avatar) {
    this._name = name;
    this._avatar = avatar;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string | null} The participant avatar, or null if there isn't one.
   */
  public get avatar(): Avatar | null {
    return this._avatar || null;
  }

  /**
   * Gets whether the participant has an avatar or not.
   * @returns {boolean} Whether the participant has an avatar or not.
   */
  @computed public get hasAvatar(): boolean {
    return !!this._avatar;
  }

  /**
   * Gets the capitalized name of the participant.
   * @returns {string} The capitalized name.
   */
  @computed public get properName(): string {
    return this._name.replace(/\w\S*/g, (subText: string) => {
      return subText.charAt(0).toUpperCase() + subText.substr(1).toLowerCase();
    });
  }

  private isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  private isSameAvatar = (avatar1?: Avatar, avatar2?: Avatar): boolean =>
    avatar1 && avatar2
      ? avatar1.equals(avatar2)
      : !avatar1 && !avatar2
      ? true
      : false;

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @returns {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this.isSameName(this._name, otherParticipant._name) &&
    this.isSameAvatar(this._avatar, otherParticipant._avatar);
}
