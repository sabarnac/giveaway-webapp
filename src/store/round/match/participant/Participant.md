Usage:

```tsx
import Avatar from "./store/round/match/participant/Avatar";

const participant: Participant = new Participant(
  "FOO BAR" /*, {url:"url",altText:"alt-text"} */
);
const participantName: string = participant.name; // "FOO BAR"
const participantProperName: string = participant.properName; // "Foo Bar"
const participantAvatar: Avatar = participant.avatar; // Avatar("url","alt-text")
```
