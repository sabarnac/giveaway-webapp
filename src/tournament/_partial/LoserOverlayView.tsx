import LoserOverlayView from "./LoserOverlayView.react";

export default LoserOverlayView;
