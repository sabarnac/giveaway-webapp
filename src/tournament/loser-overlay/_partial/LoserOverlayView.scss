.loser-overlay {
  border: 1px dashed #000000;
  position: relative;
  background: #ffffff;
  max-width: 80vw;
  min-width: 50vw;
  max-height: 90vh;
  overflow-y: auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 4rem 4rem 2rem 4rem;
}
