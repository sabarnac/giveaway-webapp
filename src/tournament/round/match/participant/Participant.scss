.participant {
    background: #ffffff;
    &--invert {
        background: #000000;
        color: #ffffff;
    }
}