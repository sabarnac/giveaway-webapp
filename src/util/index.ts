export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
