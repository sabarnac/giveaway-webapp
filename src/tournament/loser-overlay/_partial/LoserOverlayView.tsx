import React, { useEffect } from "react";
import { Observer } from "mobx-react";
import "./LoserOverlayView.scss";
import classNames from "classnames";
import Participant from "../../../store/round/match/participant/Participant";
import {
  AnimationStateHookResult,
  useAnimationState,
  runOnPredicate,
} from "../../../util/index";
import LoserInfo from "./LoserInfo";

/**
 * Properties of the loser overlay React component.
 */
interface LoserOverlayViewProps {
  /** CSS class */
  className: string;
  /** The losers of a round. */
  losers: Participant[];
  /** Action to call when the overlay has finished showing the list of losers. */
  onViewComplete: () => void;
}

/**
 * React component for the loser overlay.
 */
export default (props: LoserOverlayViewProps): JSX.Element => {
  const [
    currentState,
    updateState,
  ]: AnimationStateHookResult = useAnimationState();

  useEffect(
    runOnPredicate(currentState === props.losers.length, props.onViewComplete),
  );

  return (
    <Observer>
      {() => (
        <div className={classNames(props.className)}>
          <h2>Losers</h2>
          {props.losers.map(
            (participant: Participant, index: number): JSX.Element => (
              <LoserInfo
                key={participant.name}
                className={props.className}
                show={currentState === index}
                participant={participant}
                onInfoComplete={updateState}
              />
            ),
          )}
        </div>
      )}
    </Observer>
  );
};
