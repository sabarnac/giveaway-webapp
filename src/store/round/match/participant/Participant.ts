import { observable, computed } from "mobx";
import { createAvatarImage } from "../../../../util";
import Avatar from "./Avatar";
const inflect = require("i")();

interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  /** The ID of the participant. */
  @observable private _id: string;
  /** The name of the participant. */
  @observable private _name: string;
  /** The avatar of the participant. */
  @observable private _avatar: Avatar;
  /** The weight of the participant. */
  @observable private _weight: number;

  /** A counter for generating a unique ID for the match. */
  private static counter: number = 1;

  /**
   * Gets the avatar of the participant if present, or creates a random one,
   * @param  {string} name The name of the participant.
   * @param  {AvatarJson} avatar The avatar details of the participant, if available.
   * @return {Avatar} The generated avatar of the participant.
   */
  private _getOrCreateAvatar = (name: string, avatar?: AvatarJson): Avatar => {
    return avatar
      ? new Avatar(avatar.url, avatar.altText)
      : new Avatar(createAvatarImage(name), name);
  };

  public constructor(name: string, avatar?: AvatarJson, weight: number = 1) {
    this._id = `${Participant.counter++}`;
    this._name = name;
    this._avatar = this._getOrCreateAvatar(name, avatar);
    this._weight = weight;
  }

  /**
   * Get the participant ID.
   * @return {string} The unique ID of the participant.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  @computed public get name(): string {
    return this._name;
  }

  /**
   * Gets the capitalized name of the participant.
   * @return {string} The capitalized name.
   */
  @computed public get properName(): string {
    return inflect.titleize(this._name);
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string} The participant avatar.
   */
  @computed public get avatar(): Avatar {
    return this._avatar;
  }

  /**
   * Gets the participant weight.
   * @return {number} The participant weight.
   */
  @computed public get weight(): number {
    return this._weight;
  }

  /**
   * Returns whether the two names are equal.
   * @param {string} name1 The first name to compare.
   * @param {string} name2 The second name to compare.
   * @return {boolean} Whether the two names are equal.
   */
  private _isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  /**
   * Returns whether the two avatars are equal.
   * @param {Avatar} avatar1 The first avatar to compare.
   * @param {Avatar} avatar2 The second avatar to compare.
   * @return {boolean} Whether the two avatars are equal.
   */
  private _isSameAvatar = (avatar1: Avatar, avatar2: Avatar): boolean =>
    avatar1.equals(avatar2);

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @return {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this._isSameName(this._name, otherParticipant._name) &&
    this._isSameAvatar(this._avatar, otherParticipant._avatar);
}
