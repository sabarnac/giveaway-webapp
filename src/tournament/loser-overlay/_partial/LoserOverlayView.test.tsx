import React from "react";
import LoserOverlayView from "./LoserOverlayView.react";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyParticipant } from "../../../util/test";
import Participant from "../../../store/round/match/participant/Participant";

describe("LoserOverlayView Component.", () => {
  it("Matches snapshot.", () => {
    const losers: Participant[] = [
      createDummyParticipant(1),
      createDummyParticipant(2),
      createDummyParticipant(3),
      createDummyParticipant(4),
    ];
    const component: ReactTestRenderer = create(
      <LoserOverlayView
        className="foobar"
        losers={losers}
        onViewComplete={() => {}}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
