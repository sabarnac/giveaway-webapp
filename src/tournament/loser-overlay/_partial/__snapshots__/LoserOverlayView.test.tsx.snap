// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoserOverlayView Component. Matches snapshot. 1`] = `
<div
  className="foobar"
>
  <h2>
    Losers
  </h2>
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__loser--entering",
        "enterDone": "foobar__loser--entered",
        "exit": "",
        "exitActive": "foobar__loser--exiting",
        "exitDone": "foobar__loser--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__loser"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-card"
      >
        <div
          className="participant-card__avatar"
        >
          <img
            alt="foobar-1"
            src="foobar-1"
          />
        </div>
        <div
          className="participant-card__name"
        >
          <strong>
            Fozbaz 1
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__loser--entering",
        "enterDone": "foobar__loser--entered",
        "exit": "",
        "exitActive": "foobar__loser--exiting",
        "exitDone": "foobar__loser--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__loser"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-card"
      >
        <div
          className="participant-card__avatar"
        >
          <img
            alt="foobar-2"
            src="foobar-2"
          />
        </div>
        <div
          className="participant-card__name"
        >
          <strong>
            Fozbaz 2
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__loser--entering",
        "enterDone": "foobar__loser--entered",
        "exit": "",
        "exitActive": "foobar__loser--exiting",
        "exitDone": "foobar__loser--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__loser"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-card"
      >
        <div
          className="participant-card__avatar"
        >
          <img
            alt="foobar-3"
            src="foobar-3"
          />
        </div>
        <div
          className="participant-card__name"
        >
          <strong>
            Fozbaz 3
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__loser--entering",
        "enterDone": "foobar__loser--entered",
        "exit": "",
        "exitActive": "foobar__loser--exiting",
        "exitDone": "foobar__loser--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__loser"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-card"
      >
        <div
          className="participant-card__avatar"
        >
          <img
            alt="foobar-4"
            src="foobar-4"
          />
        </div>
        <div
          className="participant-card__name"
        >
          <strong>
            Fozbaz 4
          </strong>
        </div>
      </div>
    </div>
  </div>
</div>
`;
